### DISCUSSION TITLE
Automatically infer the return type of a trpc query using prisma

### DISCUSSION THREAD

**Discussion by TheLexoPlexx** [trpc/trpc#4475] [Q&A]
Hi,

I just added tRPC to my Nextjs and Prisma and was quite amazed about the abilites. The reason I added this was, that client and server components in nextjs being nested into each other got quite complicated and I wanted to dive into a different approach.

I set up trpc through the starter kit and with a few smaller errors, worked out generally quite well, which is really cool.

However, when using a simple query such as getting all users from a database, I noticed that the return type is not automatically being set. I looked for that and found `.output()` which does what I want, but setting this for more complicated prisma-requests using include and select gets complicated fast as well.

What I found next was the [prisma-trpc-generator](https://github.com/omar-dulaimi/prisma-trpc-generator) which is a really nice tool as well but sadly didn't work out very well for unknown reasons.

All these hours of digging and debugging got me back to the original question: How can I easily set the return type of a specific function? The main selling point of this was the E2E in the first place and noticing that half of that isn't included is sort of disappointing.

Or I might be doing something wrong.

Any help is appreciated.

(Yes, I already found the [prisma-zod-generator](https://github.com/omar-dulaimi/prisma-zod-generator) as well, might try that after this one.

Might as well remove trpc from the stack again and just go with server actions.

Any help is appreciated.

**Comment by TheLexoPlexx** [comment]
So apparently, this is possible through inferring types: https://trpc.io/docs/client/infer-types

But makes me end up with something like this:
```ts
const users: RouterOutputs["users"]["getUsersForChangeUserDialog"] | undefined = trpc.users.getUsersForChangeUserDialog.useQuery().data
```

And this may get annoying with more and more requests, but I've seen worse code. I'll leave this open for now until someone finds a more elegant way.

**Reply by linkfang** [reply]
In theory, (and also in my case), the type should be auto inferred. 

So, I don't need to manually assign the type using `RouterOutputs`. 
I will only need to use it when I just need the type of it and assign it to something else other than the query data.

Maybe start a new test project and do a simple minimal data model and try there?

**Reply by linkfang** [reply]
See screenshot below. I am also using [prisma-trpc-generator](https://github.com/omar-dulaimi/prisma-trpc-generator)
![image](https://github.com/trpc/trpc/assets/43075439/813acda3-ab93-45a2-a9fc-85226d9e035d)

**Reply by TheLexoPlexx** [reply]
As far as I remember, I had a similar result in the end as well, but I don't remember anymore why. I also ditched trpc when server actions dropped.

**Reply by ackvf** [reply]
The query/mutation indeed infers the correct return type from a prisma request, but we need the type itself as well.
Is there a generator that would output all Input/Output types for all queries and mutations, without requiring us to do this?

```ts
import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server"
import type { AppRouter } from "@/server/trpc/routers/_app"

type RouterInput = inferRouterInputs<AppRouter>
type RouterOutput = inferRouterOutputs<AppRouter>

export type UserListInput = RouterInput["user"]["list"]
export type UserListOutput = RouterOutput["user"]["list"]
// repeat for ALL routers
```
In the screenshots below I can confirm that the Prisma correctly infers the parameters from code, which in turn gets correctly picked up by trpc:

<img width="669" height="407" alt="image" src="https://github.com/user-attachments/assets/e8d96877-6233-48f1-a1e9-72a6ea048537" />

And it indeed gets all the way to the resulting data.
<img width="475" height="110" alt="image" src="https://github.com/user-attachments/assets/5510e01e-3a6e-4326-add4-c092c147e0ce" />

But then I can't do something like this

```
const [user, setUser] = useState<UserResponseType>() // <- don't have UserResponseType
```
